### PR TITLE
VACMS-3436: Improves link experience on help center pages.

### PR DESCRIPTION
**Pull request by ndouglas**
## Description

See #3436.

Now updated to address the actual AC's 😎 

## Testing done
Forthcoming.  I'll add a Behat test or two on Monday.

## Screenshots
Screen Shot 2020-12-07 at 10 24 58 AM
Screen Shot 2020-12-07 at 10 29 06 AM
Screen Shot 2020-12-07 at 10 51 08 AM

## QA steps

As a user with permission to edit content:
1. View a [`documentation_page` node with a `downloadable_file` paragraph containing a document](http://va-gov-cms.lndo.site/node/9421).
1. Then validate AC's:
    - [x] "Link to file or video" paragraph display contains only a link with the little download icon
    - [x] Link markup looks like: `<a class="downloadable-file-link downloadable-file-link--document" target="_blank" href="{{url}}" download="{{url}}" aria-label="Download">Download slide deck for Session 1 - CMS Orientation (PDF)</a>`
1. View a [`documentation_page` node with a `downloadable_file` paragraph containing an image](http://va-gov-cms.lndo.site/node/7166).
1. Then validate AC's:
    - [x] "Link to file or video" paragraph display contains only a link with the little download icon
    - [x] Link markup looks like: `<a class="downloadable-file-link downloadable-file-link--image" target="_blank" href="{{url}}" download="{{url}}" aria-label="Download">Download Salem Visitor's Guide (PNG)</a>`
    - this will fail until the content is updated to conform to the new guidelines
1. View a `documentation_page` node with a `downloadable_file` paragraph containing a video, or add one to [an existing node](http://va-gov-cms.lndo.site/node/9422).
1. Then validate AC's:
    - [x] "Link to file or video" paragraph display contains only a link with the little video icon
    - [x] Link markup looks like: `<a class="downloadable-file-link downloadable-file-link--video" target="_blank" href="{{url}}" download="{{url}}" aria-label="Download">This is a test video.</a>`
1. View a [**non**-`documentation_page` node with a `downloadable_file` paragraph containing a video](http://va-gov-cms.lndo.site/node/6172).
1. Then validate AC's:
    - [x] "Link to file or video" paragraph display contains only a link with the little video icon
    - [x] Link markup looks like: `<a class="downloadable-file-link downloadable-file-link--video" target="_blank" href="{{url}}" download="{{url}}" aria-label="Download">Go to video</a>`

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [x] `Core Application Team`
- [ ] `Product Support Team`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change? 
- [ ] Yes, and it's written in issue ____ and queued for publication. 
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [ ] Yes, but it hasn't yet been written 
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [x] No announcement is needed for this code change. 
  - [x] Merge & carry on unburdened by announcements

### CODE REVIEW

File: config/sync/field.field.paragraph.downloadable_file.field_media.yml [hunk 1]
```diff
@@ -6,13 +6,18 @@ dependencies:
...
+third_party_settings:
+  entity_reference_validators:
+    circular_reference: false
+    duplicate_reference: false
 id: paragraph.downloadable_file.field_media
 field_name: field_media
 entity_type: paragraph
 bundle: downloadable_file
-label: 'Link to a file or video'
+label: 'Link to a file'
```
Comment (oksana-c) [inline config/sync/field.field.paragraph.downloadable_file.field_media.yml:20]: per ACs - paragraph config should not be changing in this PR

File: config/sync/paragraphs.paragraphs_type.downloadable_file.yml [hunk 2]
```diff
@@ -7,9 +7,9 @@ dependencies:
 third_party_settings:
   paragraphs_browser:
     image_path: themes/custom/vagovadmin/images/screenshots/download-map.jpg
-    description: "Allowed file types: \r\nFor images: png, gif, jpg, jpeg \r\nFor files: pdf, doc, docx, ppt, xlsx (pdf is preferable)\r\nVideo support coming soon!"
 id: downloadable_file
-label: 'Link to file or video'
+label: 'Link to file'
```
Comment (oksana-c) [inline config/sync/paragraphs.paragraphs_type.downloadable_file.yml:11]: same - not changing paragraph config

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module [hunk 3]
```diff
@@ -1346,6 +1348,53 @@ function va_gov_backend_toolbar() {
...
+        $base_class,
+        $base_class . '--' . $media_bundle,
+      ];
+      $media_object = [];
+      $media_object['link_title'] = $title;
+      $media_object['attributes'] = new Attribute([]);
+      $media_object['attributes']->addClass($classes);
+      if ($media_bundle === 'file' || $media_bundle === 'image') {
+        $fid = $media_entity->getSource()->getSourceFieldValue($media_entity);
+        if (!empty($fid)) {
```
Comment (oksana-c) [inline docroot/modules/custom/va_gov_backend/va_gov_backend.module:1385]: wondering if we could skip some of these calls/preprocessing by setting the rendered entity format to "Download link"

Screen Shot 2020-12-08 at 6 02 22 AM

We also need to have a file mime type displayed in caps for files and images only, not for videos

---

We're also missing File links for Media:File
Screen Shot 2020-12-08 at 6 09 21 AM

File: docroot/themes/custom/vagovadmin/templates/paragraph/paragraph--downloadable-file--documentation-page.html.twig [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * @see template_preprocess_paragraph()
+ *
+ * @ingroup themeable
+ */
+#}
+{% extends "paragraph.html.twig" %}
+{% block content %}
+  {% for media_object in field_media %}
+    <span {{ media_object.attributes }}>
```
Comment (oksana-c) [inline docroot/themes/custom/vagovadmin/templates/paragraph/paragraph--downloadable-file--documentation-page.html.twig:44]: looks to me that we could omit this extra span and move `{{ media_object.attributes }}` to the link markup. The icon then would be set in `&::before` selector

**Review by oksana-c — CHANGES_REQUESTED** [review]
I know this is in Draft still, but figured I'll note a few things:

1. The download link is empty for Media:File (see screenshot in comments)
2. The file extension is missing for Media:File and Media:Image. e.g. `(PDF)`
3. Video remains as an option for "Link to file or video" paragraph. Let's revert paragraph config that unsets video and changes paragraph's name
4. NEW: We'll need distinct template logic for `documentation_page` content type and other content types when it comes to rendering Media:Video (see comment on the template file)

Let's zoom if it's easier to run thru all these IRL...  just called zoom IRL 🤔

File: config/sync/field.field.paragraph.downloadable_file.field_media.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ndouglas) [reply config/sync/field.field.paragraph.downloadable_file.field_media.yml]: Yeah, I thought I changed that back but must've not exported config 🤦

**Comment by ndouglas** [comment]
The fourth Behat test will fail ATM because there're no `downloadable_file` paragraphs with videos that aren't attached to `documentation_page` nodes.  The modal dialog seems to break BeHat's ability to gracefully test this sort of operation, so I'm going to have to fiddle with possible solutions tomorrow.

**Comment by ndouglas** [comment]
@oksana-c @acrollet 

Do you have any thoughts about testing this fourth situation (non-`documentation_page` node with `downloadable_file` video)?  

I see ~three~ ~four~ five options:

1) use [this](https://github.com/dennisinteractive/behat-drupal-paragraphs) to create nodes with paragraphs with specific structure and content.  However, it hasn't been updated in a couple years and it's currently not installable with Composer (haven't dug into why, but seems like a bad sign).  This would probably be very useful in other tests.

2) create dummy content on prod so the DB has something to test against.  I dislike this and I don't think it's scalable, but it's fast.

3) Try to get the test working by creating a node via pushing buttons and such in BeHat.  This was the approach I used yesterday and it's not great -- the modal dialog inserts a paragraph (of the wrong type) as soon as the "Add Content block" button is pushed, which I think is because of the Goutte driver and the way modal dialogs are done in Drupal.  This is a lousy and fragile way to do things, overall, and it's a major PITA to write tests this way.

4) Fork or rewrite `dennisinteractive/behat-drupal-paragraphs` for our own use, since it looks abandoned.

5) Write the test(s) in something other than BeHat.

**Comment by acrollet** [comment]
My vote is to look into whether we could use cypress for our e2e testing now that we're close to having tugboat up. It's a much nicer framework to use and the tests run on a full browser, so it's much closer to the actual environment. I have a working branch with cypress-axe that I'll dig up when I'm online.

**Comment by acrollet** [comment]
Here's the branch: https://github.com/acrollet/va.gov-cms/tree/VACMS-3289-implement-cypress-axe

**Comment by acrollet** [comment]
ha @ndouglas I was about to make a separate PR for the stylelint issues - I'm guessing it was something that got merged around the same time as the stylelint PR.

**Comment by ndouglas** [comment]
nah, it was my crap (s)css

**Comment by acrollet** [comment]
haha meh it's a matter of opinion... 😄

**Comment by ndouglas** [comment]
@acrollet I dig Cypress, and would way rather write JS than Gherkin or whatever, but I guess it's a bit much to roll into this PR.  

I'd love to make it a top priority, though, because it's pretty aggravating to not have a clean way (AFAICT) of writing tests for this stuff.

**Comment by acrollet** [comment]
@ndouglas we'd certainly have to do it over time, but I'd like to have us consider writing new e2e tests in cypress. [this](https://github.com/TheBrainFamily/cypress-cucumber-preprocessor) could potentially help with migration...

I think we have an open issue about e2e testing tools somewhere.

**Comment by ndouglas** [comment]
a'ight, I'ma do it -- I'm not thinking up any faster solutions tbh, and I'm on the clock 🕞

**Comment by acrollet** [comment]
Sorry, I should have been more clear that we probably need a team-level decision on this.￼ I know that Elijah didn't want to introduce Cypress until after tugboat is up and running. The solution for this issue might be to test manually and add a follow on to automate the test. cc oksana-c

**Comment by ndouglas** [comment]
too late now rollett, I'm a rootin'-tootin' [ranagazoo](https://www.youtube.com/watch?v=qLrnkK2YEcE)

File: docroot/themes/custom/vagovadmin/templates/paragraph/paragraph--downloadable-file--documentation-page.html.twig [hunk 5]
```diff
@@ -0,0 +1,51 @@
+{#
+/**
+ * @file
+ * Default theme implementation to display a paragraph.
```
Comment (oksana-c) [inline docroot/themes/custom/vagovadmin/templates/paragraph/paragraph--downloadable-file--documentation-page.html.twig:4]:
```suggestion
 * VAGOVADMIN theme implementation to display a `downloadable_file` paragraph 
 on `documentation_page` nodes.
```

File: docroot/themes/custom/vagovadmin/templates/paragraph/paragraph--downloadable-file.html.twig [hunk 6]
```diff
@@ -0,0 +1,51 @@
+{#
+/**
+ * @file
+ * Default theme implementation to display a paragraph.
```
Comment (oksana-c) [inline docroot/themes/custom/vagovadmin/templates/paragraph/paragraph--downloadable-file.html.twig:4]:
```suggestion
 * VAGOVADMIN theme implementation to display a `downloadable_file` paragraph.
```

File: docroot/themes/custom/vagovadmin/assets/scss/styles.scss [hunk 7]
```diff
@@ -165,6 +165,28 @@
...
 
+.downloadable-file-link::before {
+  background-repeat: no-repeat;
+  background-size: 16px 16px;
+  content: '';
+  display: inline-block;
+  height: 16px;
+  padding-right: 4px;
+  width: 16px;
+}
```
Comment (oksana-c) [inline docroot/themes/custom/vagovadmin/assets/scss/styles.scss:176]:
```suggestion
.downloadable-file-link {
  padding-left: 20px;
  position: relative;

  &::before {
    background-repeat: no-repeat;
    background-size: 16px 16px;
    content: '';
    display: inline-block;
    height: 16px;
    padding-right: 4px;
    width: 16px;
  }
}
```

couldn't help to make it pixel perfect..

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module [hunk 8]
```diff
@@ -1346,6 +1348,53 @@ function va_gov_backend_toolbar() {
...
+    $suggestions[] = 'paragraph__' . $paragraph->bundle() . '__' . $parent->bundle();
+  }
+}
+
+/**
+ * Implements template_preprocess_paragraph().
+ */
+function va_gov_backend_preprocess_paragraph(&$variables) {
+  $paragraph = $variables['paragraph'];
+  $parent_bundle = $paragraph->getParentEntity()->bundle();
```
Comment (oksana-c) [inline docroot/modules/custom/va_gov_backend/va_gov_backend.module:1367]:
```suggestion
```

unused variable

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module [hunk 9]
```diff
@@ -22,6 +22,8 @@ use Drupal\va_gov_build_trigger\Form\PreviewForm;
 use Symfony\Cmf\Component\Routing\RouteObjectInterface;
 use Symfony\Component\HttpFoundation\RedirectResponse;
 use Drupal\Component\Utility\Crypt;
+use Drupal\file\Entity\File;
+use Drupal\Core\Template\Attribute;
```
Comment (oksana-c) [inline docroot/modules/custom/va_gov_backend/va_gov_backend.module:26]: A-Z sort on use statements would be nice

File: config/sync/paragraphs.paragraphs_type.downloadable_file.yml [hunk 10]
```diff
@@ -7,9 +7,9 @@ dependencies:
 third_party_settings:
   paragraphs_browser:
     image_path: themes/custom/vagovadmin/images/screenshots/download-map.jpg
-    description: "Allowed file types: \r\nFor images: png, gif, jpg, jpeg \r\nFor files: pdf, doc, docx, ppt, xlsx (pdf is preferable)\r\nVideo support coming soon!"
```
Comment (oksana-c) [inline config/sync/paragraphs.paragraphs_type.downloadable_file.yml:10]: let's bring this one back 

`description: "Allowed file types: \r\nFor images: png, gif, jpg, jpeg \r\nFor files: pdf, doc, docx, ppt, xlsx (pdf is preferable)\r\nVideos can be selected from the library."`

**Review by oksana-c — COMMENTED** [review]
there ain't no stoppin' ya Nate! 

I agree w/ @acrollet, we need to hold off on Cypress. Glad you both got to experiment w/ it ahead of time though 👍  
Let's remove video tests, I validated the ACs manually in DevShop, so once vid tests are removed and build is passing - let's get this one merged.
Added a few inline suggestions, some of which you can commit right in github UI.
Links look great 💪

File: docroot/themes/custom/vagovadmin/assets/scss/styles.scss [hunk 11]
```diff
@@ -165,6 +165,33 @@
...
+  position: relative;
+
+  &::before {
+    background-repeat: no-repeat;
+    background-size: 16px 16px;
+    content: '';
+    display: inline-block;
+    height: 16px;
+    padding-right: 4px;
+    width: 16px;
```
Comment (oksana-c) [inline docroot/themes/custom/vagovadmin/assets/scss/styles.scss:179]:
```suggestion
    background-repeat: no-repeat;
    background-size: contain;
    content: '';
    display: block;
    height: 16px;
    position: absolute;
    top: 2px;
    width: 16px;
```

**Review by oksana-c — APPROVED** [review]

**Comment by oksana-c** [comment]
✅ ready to merge once tests pass.

@rachel-kauff , this PR once deployed will add file extensions to link automatically, so you'll need to update some link titles in CMS Help Center that have file ext intentionally added. example on /help/session-1-cms-orientation page 
Screen Shot 2020-12-10 at 6 23 04 PM

**Review by oksana-c — APPROVED** [review]

**Pull request merged**